Implement MMVv1 instance metrics

saurvs opened the pull request:
Apart from implementing instance metrics, this also includes some cleanup/refactoring of singleton metric writing in order to improve code reuse.

saurvs committed:
Implement MMVv1 instance metrics

ryandoyle commented on src/client/metric.rs:
```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
+            im.indom = (self.running_hasher.finish() as u32) & ((1 << INDOM_BIT_LEN) - 1);
+
+            let metric = Metric::new(
+                &instance,
```
Maybe I'm reading this wrong but will this make the metric name the instance name? A group of instances should belong to a single metric name.

suyash replied:
I believe the idea here is that the Instance Metric contains inside it Singleton Metrics that have the same name as the instance name. This way, we can maintain unique name, shorthelp and longhelp, but duplicate semantics, unit and indom for each instance

ryandoyle commented on src/client/metric.rs:
```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
+    use super::Client;
+
+    let mut cache = InstanceMetric::new(
+        Semantics::Discrete,
```
Like the previous comment, how is the metric name defined here?

ryandoyle reviewed (commented):
Looking good - can we also add an example in the `examples/` folder?

suyash commented on src/client/metric.rs:
```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
 }
 
+pub struct InstanceMetric<T> {
+    pub (super) metrics: HashMap<String, Metric<T>>,
```
I see that your idea of an instance metric is multiple Singleton Metrics wrapped inside a Combined metric type with instances, but doesn't that seem an overkill, especially since all other attributes (semantics, unit) will be the same for all internal metrics. Maybe we can have a `metricData` type managing singleton data used by both Singleton and Instance Metrics.

suyash replied:
Although, on a second look the amount of duplicated data seems small enough, so that might be overkill

suyash commented on src/client/metric.rs:
```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
+                longhelp: longhelp_text.to_owned(),
+                indom: 0
+            },
+            running_hasher: DefaultHasher::new()
```
Any reason in particular to use a `Hasher` and not generate completely random values, Also, please document all public types and methods

suyash commented on src/client/mod.rs:
```diff
@@ -216,12 +247,29 @@ impl Client {
     }
 
-    pub fn begin(&mut self, n_metrics: u64) -> io::Result<&mut Client> {
+    pub fn begin(&mut self, n_indoms: u64, n_instances: u64, n_metrics: u64) -> io::Result<&mut Client> {
```
I know it isn't possible in the current implementation of the writer to do this as we write tocs ahead of time, but lets explore not requiring the user to specify the number of instances and indoms beforehand.

suyash replied:
Also, since last year's program ended, I have thought about multiple times to make indom a part of the implementation and not expose the idea to the user at all, and put all instances of a metric in an autogenerated indom

suyash reviewed (commented):
This seems to be the quickest way of getting instance metrics to work, and is extremely clean and simple which is good. With a few tweaks, I think we can merge this. Again, Good work!

suyash commented on src/client/metric.rs:
```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
+
+impl<T: MetricType + Clone> InstanceMetric<T> {
+    pub fn new(
+        sem: Semantics,
```
An InstanceMetric does have its own name, and values are represented as `name[instance1]`, `name[instance2]` etc.

suyash commented on src/client/metric.rs:
```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
+                unit: unit,
+                shorthelp: shorthelp_text.to_owned(),
+                longhelp: longhelp_text.to_owned(),
+                indom: 0
```
indom is the value that will uniquely identify a group of metrics, should probably be generated and set to a random value instead of 0, otherwise for multiple instance metrics, won't their instances be classified under the same indom?

saurvs replied:
`mmvdump` ends up displaying zero indom values as `(no indom)` so I think we should be safe.

saurvs commented:
@ryandoyle @suyash After trying to write the ACME example using the API in this PR, I realised what I've implemented here is quite different from `libpcp_mmv` or Speed.

Here, each `InstanceMetric` ends up defining it's own instance domain, and creates a metric block and value block for each instance block. However, in the C library, each instance from an instance domain can actually be linked to multiple metric and value blocks. This is why for Indom{Anvils, Rockets} and Metrics{Count, Time}, we get 4 distinct value blocks. 

If we tried to implement ACME using this PR using InstanceMetric{Count}{Anvils, Rockets} and  InstanceMetric{Time}{Anvils, Rockets}, we'd get 2 identical indom blocks, 2 pairs of identical instance blocks, 4 metrics and 4 value blocks, when it really should be 1 indom, 2 instance, 2 metric and 4 value blocks.

I wanted to try out the C model, hence I implemented a separate fork of this PR in another branch (https://github.com/performancecopilot/hornet/tree/indom) where the Indom concept is explicit, and the ACME example is implemented similar to the C version (https://github.com/performancecopilot/hornet/blob/indom/examples/acme.rs).

I think the basic question is, should an `InstanceMetric` represent **one** metric associated with multiple instances, or should it represent multiple instances each with a metric of it's own?

suyash commented:
> I think the basic question is, should an InstanceMetric represent one metric associated with multiple instances,

Yes

> or should it represent multiple instances each with a metric of it's own?

That can be how its done internally. One of the things I like about this is that it tries to abstract out the common stuff between a metric and an instance

> If we tried to implement ACME using this PR using InstanceMetric{Count}{Anvils, Rockets} and InstanceMetric{Time}{Anvils, Rockets}, we'd get 2 identical indom blocks, 2 pairs of identical instance blocks, 4 metrics and 4 value blocks, when it really should be 1 indom, 2 instance, 2 metric and 4 value blocks.

I now realize that this is precisely the reason why we need a way to __create indoms__ in the API, and then allow an instance metric to set its indom. Although we can allow the API to __create instances__ and add them and every time the instances are changed, we generate a random indom id internally. Again, I really like the idea here to have the "indom type" as simply an integer, but it provides no way to track the instances in an indom.

Can you open a separate PR with the other branch to open it up for review? Just do that every time, its OK to have PRs just for review.

suyash commented:
I now see that the `indom` branch depends on this branch. Just fast-forward this branch to that commit, since that is probably the way we want to go to.

saurvs commented:
Closing in favour of https://github.com/performancecopilot/hornet/pull/23
